omniscidb/QueryEngine/ArrowResultSetConverter.cpp

```diff
@@ -786,18 +786,33 @@ size_t convert_rowwise(
     std::vector<std::shared_ptr<std::vector<bool>>>& null_bitmap_seg,
     const std::vector<bool>& non_lazy_cols,
     const size_t start_entry,
-    const size_t end_entry) {
+    const size_t end_entry,
+    bool is_truncated = false) {
   const auto col_count = results->colCount();
   CHECK_EQ(value_seg.size(), col_count);
   CHECK_EQ(null_bitmap_seg.size(), col_count);
   const auto local_entry_count = end_entry - start_entry;
   size_t seg_row_count = 0;
+  size_t limit = 0;
+  size_t offset = end_entry;
+  if (is_truncated) {
+    limit = results->getLimit();
+    offset = results->getOffset();
+  }
   for (size_t i = start_entry; i < end_entry; ++i) {
     auto row = results->getRowAtNoTranslations(i, non_lazy_cols);
     if (row.empty()) {
       continue;
     }
     ++seg_row_count;
+
+    if (is_truncated && seg_row_count <= offset) {
+      continue;
+    }
+    if (is_truncated && seg_row_count > offset + limit) {
```

Comment:
We don't need to look for the next valid entry if converted enough rows. So please move this exit condition to the very beginning of the loop.

Reply:
Agree,